flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java

```diff
@@ -306,22 +331,88 @@ void deliverOperatorEventToCoordinator(
                 operatorId, request);
     }
 
+    /** Transition to different state when failure occurs. Stays in the same state by default. */
+    abstract void onFailure(Throwable cause);
+
+    /**
+     * Transition to different state when the execution graph reaches a globally terminal state.
+     *
+     * @param globallyTerminalState globally terminal state which the execution graph reached
+     */
+    abstract void onGloballyTerminalState(JobStatus globallyTerminalState);
+
+    @Override
+    public void handleGlobalFailure(Throwable cause) {
+        failureCollection.add(new GlobalFailure(cause));
+        onFailure(cause);
+    }
+
     /**
      * Updates the execution graph with the given task execution state transition.
      *
      * @param taskExecutionStateTransition taskExecutionStateTransition to update the ExecutionGraph
      *     with
      * @return {@code true} if the update was successful; otherwise {@code false}
      */
-    abstract boolean updateTaskExecutionState(
-            TaskExecutionStateTransition taskExecutionStateTransition);
+    boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionStateTransition) {
+        if (taskExecutionStateTransition.getExecutionState() != ExecutionState.FAILED) {
+            return getExecutionGraph().updateState(taskExecutionStateTransition);
+        }
 
-    /**
-     * Callback which is called once the execution graph reaches a globally terminal state.
-     *
-     * @param globallyTerminalState globally terminal state which the execution graph reached
-     */
-    abstract void onGloballyTerminalState(JobStatus globallyTerminalState);
+        // We need to collect the ExecutionVertexID before updating the state, because the Execution
+        // is de-registered afterwards.
+        // We need to use an optional here, because this method can be called even after the
+        // Execution is de-registered.
+        Optional<ExecutionVertexID> idOpt =
+                executionGraph.findExecutionVertexId(taskExecutionStateTransition.getID());
+        final boolean successfulUpdate =
+                getExecutionGraph().updateState(taskExecutionStateTransition);
+        if (!successfulUpdate) {
+            return false;
+        }
+
+        Throwable cause = extractErrorOrUseDefault(taskExecutionStateTransition);
+
+        checkState(idOpt.isPresent());
+        ExecutionVertexID id = idOpt.get();
+
+        if (getNonEmptyExecution(id).getFailureInfo().isPresent()) {
+            failureCollection.add(new LocalFailure(cause, id));
+        }
+        onFailure(cause);
+        return true;
```

Comment:
It seems you've been faster to fix this :) My suggestion would have been something along this lines:

```
    boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionStateTransition) {
        final Optional<ExecutionVertexID> maybeExecutionVertexId =
                executionGraph.findExecutionVertexId(taskExecutionStateTransition.getID());
        final boolean successfulUpdate =
                getExecutionGraph().updateState(taskExecutionStateTransition);
        if (successfulUpdate) {
            // We're sure that the executionVertexId has been found, because we've been able to
            // update the execution graph.
            final ExecutionVertexID executionVertexId =
                    maybeExecutionVertexId.orElseThrow(NoSuchElementException::new);
            final ExecutionState desiredState = taskExecutionStateTransition.getExecutionState();
            final ExecutionState newState =
                    executionGraph
                            .findExecution(executionVertexId)
                            .map(Execution::getState)
                            .orElseThrow(NoSuchElementException::new);
            // We only want a notification for the actual transition into the FAILED state.
            if (desiredState == ExecutionState.FAILED && desiredState == newState) {
                final Throwable cause = extractErrorOrUseDefault(taskExecutionStateTransition);
                if (getNonEmptyExecution(executionVertexId).getFailureInfo().isPresent()) {
                    failureCollection.add(new LocalFailure(cause, executionVertexId));
                }
                onFailure(cause);
                return true;
            }
        }
        return false;
    }
```

Two notes:
- This de-duplicates the call to the state update
- More important, this also checks whether we've actually reached the desired state, to align with the behavior of the DefaultScheduler (see SchedulerBase#isNotifiable)

Reply:
Looks better than my version. I will take yours.

Just one thing. The location of the return statements changes the semantic of the method slightly from before. Is this on purpose?

Previously `desiredState = "FAILED"` and `newState = "CANCELLING"` would return `true`. Now they return `false`.

Reply:
How about, we use this:

```java
     boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionStateTransition) {
        final Optional<ExecutionVertexID> maybeExecutionVertexId =
                executionGraph.findExecutionVertexId(taskExecutionStateTransition.getID());
        if (!getExecutionGraph().updateState(taskExecutionStateTransition)) {
          return false;
        }

        // We're sure that the executionVertexId has been found, because we've been able to
        // update the execution graph.
        final ExecutionVertexID executionVertexId =
                maybeExecutionVertexId.orElseThrow(NoSuchElementException::new);
        final ExecutionState desiredState = taskExecutionStateTransition.getExecutionState();
        final ExecutionState newState =
                executionGraph
                        .findExecution(executionVertexId)
                        .map(Execution::getState)
                        .orElseThrow(NoSuchElementException::new);
        // We only want a notification for the actual transition into the FAILED state.
        if (desiredState == ExecutionState.FAILED && desiredState == newState) {
            final Throwable cause = extractErrorOrUseDefault(taskExecutionStateTransition);
            if (getNonEmptyExecution(executionVertexId).getFailureInfo().isPresent()) {
                failureCollection.add(new LocalFailure(cause, executionVertexId));
            }
            onFailure(cause);
        }
        return true;
    }
```

Reply:
2nd thought, you're right that we should still return true even though we're not actually doing any work

Then it can be basically simplified to this:

```java
        final Optional<ExecutionVertexID> maybeExecutionVertexId =
                executionGraph.findExecutionVertexId(taskExecutionStateTransition.getID());
        final boolean successfulUpdate =
                getExecutionGraph().updateState(taskExecutionStateTransition);
        if (successfulUpdate) {
            // We're sure that the executionVertexId has been found, because we've been able to
            // update the execution graph.
            final ExecutionVertexID executionVertexId =
                    maybeExecutionVertexId.orElseThrow(NoSuchElementException::new);
            final ExecutionState desiredState = taskExecutionStateTransition.getExecutionState();
            final ExecutionState newState =
                    executionGraph
                            .findExecution(executionVertexId)
                            .map(Execution::getState)
                            .orElseThrow(NoSuchElementException::new);
            // We only want a notification for the actual transition into the FAILED state.
            if (desiredState == ExecutionState.FAILED && desiredState == newState) {
                final Throwable cause = extractErrorOrUseDefault(taskExecutionStateTransition);
                if (getNonEmptyExecution(executionVertexId).getFailureInfo().isPresent()) {
                    failureCollection.add(new LocalFailure(cause, executionVertexId));
                }
                onFailure(cause);
            }
        }
        return successfulUpdate;
        ```
        
        Sorry for the confusion

Reply:
I did another slight modification of your version

Reply:
Sorry, I missed your edit above. 

This comparison `newState == desiredState` means, that the scheduler only restarts when the Task is failing. Not sure why exactly, but this fails some tests (e.g. ExecutingTest).

Reply:
I've addressed your point regarding the early returns vs nesting.

As for the `newState == desiredState`, this fails tests so I'd rather not use it.